IMPORTING LIBRARIES

In [84]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn import svm
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import GridSearchCV

DATA COLLECTION AND PROCESSING

In [2]:
#Load the dataset
data = pd.read_csv('/content/heart_disease_data.csv')

In [3]:
data.head()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   
3   56    1   1       120   236    0        1      178      0      0.8      2   
4   57    0   0       120   354    0        1      163      1      0.6      2   

   ca  thal  target  
0   0     1       1  
1   0     2       1  
2   0     2       1  
3   0     2       1  
4   0     2       1

In [4]:
data.tail()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
298   57    0   0       140   241    0        1      123      1      0.2   
299   45    1   3       110   264    0        1      132      0      1.2   
300   68    1   0       144   193    1        1      141      0      3.4   
301   57    1   0       130   131    0        1      115      1      1.2   
302   57    0   1       130   236    0        0      174      0      0.0   

     slope  ca  thal  target  
298      1   0     3       0  
299      1   0     3       0  
300      1   2     3       0  
301      1   1     3       0  
302      1   1     2       0

In [6]:
data.shape

(303, 14)

In [8]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       303 non-null    int64  
 1   sex       303 non-null    int64  
 2   cp        303 non-null    int64  
 3   trestbps  303 non-null    int64  
 4   chol      303 non-null    int64  
 5   fbs       303 non-null    int64  
 6   restecg   303 non-null    int64  
 7   thalach   303 non-null    int64  
 8   exang     303 non-null    int64  
 9   oldpeak   303 non-null    float64
 10  slope     303 non-null    int64  
 11  ca        303 non-null    int64  
 12  thal      303 non-null    int64  
 13  target    303 non-null    int64  
dtypes: float64(1), int64(13)
memory usage: 33.3 KB


In [9]:
#Check for missing values
data.isnull().sum()

age         0
sex         0
cp          0
trestbps    0
chol        0
fbs         0
restecg     0
thalach     0
exang       0
oldpeak     0
slope       0
ca          0
thal        0
target      0
dtype: int64

In [11]:
#Statistical Measures of data
data.describe()

age         sex          cp    trestbps        chol         fbs  \
count  303.000000  303.000000  303.000000  303.000000  303.000000  303.000000   
mean    54.366337    0.683168    0.966997  131.623762  246.264026    0.148515   
std      9.082101    0.466011    1.032052   17.538143   51.830751    0.356198   
min     29.000000    0.000000    0.000000   94.000000  126.000000    0.000000   
25%     47.500000    0.000000    0.000000  120.000000  211.000000    0.000000   
50%     55.000000    1.000000    1.000000  130.000000  240.000000    0.000000   
75%     61.000000    1.000000    2.000000  140.000000  274.500000    0.000000   
max     77.000000    1.000000    3.000000  200.000000  564.000000    1.000000   

          restecg     thalach       exang     oldpeak       slope          ca  \
count  303.000000  303.000000  303.000000  303.000000  303.000000  303.000000   
mean     0.528053  149.646865    0.326733    1.039604    1.399340    0.729373   
std      0.525860   22.905161    0.469794    1.161075    0.616226    1.022606   
min      0.000000   71.000000    0.000000    0.000000    0.000000    0.000000   
25%      0.000000  133.500000    0.000000    0.000000    1.000000    0.000000   
50%      1.000000  153.000000    0.000000    0.800000    1.000000    0.000000   
75%      1.000000  166.000000    1.000000    1.600000    2.000000    1.000000   
max      2.000000  202.000000    1.000000    6.200000    2.000000    4.000000   

             thal      target  
count  303.000000  303.000000  
mean     2.313531    0.544554  
std      0.612277    0.498835  
min      0.000000    0.000000  
25%      2.000000    0.000000  
50%      2.000000    1.000000  
75%      3.000000    1.000000  
max      3.000000    1.000000

In [12]:
#Checking the distribution of Target variable
data['target'].value_counts()

target
1    165
0    138
Name: count, dtype: int64

1 -> Defective Heart
2 -> Healthy Heart

Splitting Features and Target

In [14]:
x = data.drop(columns='target', axis=1)
y = data['target']

In [15]:
print(x)

     age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
0     63    1   3       145   233    1        0      150      0      2.3   
1     37    1   2       130   250    0        1      187      0      3.5   
2     41    0   1       130   204    0        0      172      0      1.4   
3     56    1   1       120   236    0        1      178      0      0.8   
4     57    0   0       120   354    0        1      163      1      0.6   
..   ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
298   57    0   0       140   241    0        1      123      1      0.2   
299   45    1   3       110   264    0        1      132      0      1.2   
300   68    1   0       144   193    1        1      141      0      3.4   
301   57    1   0       130   131    0        1      115      1      1.2   
302   57    0   1       130   236    0        0      174      0      0.0   

     slope  ca  thal  
0        0   0     1  
1        0   0     2  
2        2   0    

In [16]:
print(y)

0      1
1      1
2      1
3      1
4      1
      ..
298    0
299    0
300    0
301    0
302    0
Name: target, Length: 303, dtype: int64


Splitting Train and Test Data

In [18]:
x_train, x_test, y_train, y_test = train_test_split(x,y, test_size=0.2 , stratify=y, random_state=2)

In [20]:
x_train.shape

(242, 13)

In [21]:
x_test.shape

(61, 13)

MODEL TRAINING

In [69]:
#LOGISTIC REGRESSION

logistic_model = LogisticRegression(max_iter=500)

In [24]:
# SUPPORT VECTOR CLASSIFIER

classifier = svm.SVC(kernel = 'linear')

In [56]:
#DECISION TREE MODEL
dt_model = DecisionTreeClassifier()

In [25]:
#RANDOM FOREST

rf_model = RandomForestClassifier(n_estimators=100)

In [70]:
#Training the Logistic Regression model with Training Data
logistic_model.fit(x_train,y_train)


/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:460: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression(max_iter=500)

In [27]:
#Training the SVM model with Training Data
classifier.fit(x_train,y_train)

SVC(kernel='linear')

In [57]:
#Training the Decision Tree model with Training Data
dt_model.fit(x_train, y_train)

DecisionTreeClassifier()

In [29]:
#Training the Random Forest model with Training Data
rf_model.fit(x_train,y_train)

RandomForestClassifier()

MODEL EVALUATAION

Logistic Regressiong Accuracy Score

In [71]:
#Accuracy on Training Data
lr_x_train_prediction = logistic_model.predict(x_train)
lr_training_data_accuracy = accuracy_score(lr_x_train_prediction,y_train)

In [76]:
print("Logistic Regression Accuracy on Training Data:", lr_training_data_accuracy)

Logistic Regression Accuracy on Training Data: 0.859504132231405


In [73]:
#Accuraccy on test data
lr_x_test_prediction = logistic_model.predict(x_test)
lr_testing_data_accuracy = accuracy_score(lr_x_test_prediction,y_test)

In [77]:
print("Logistic Regression Accuracy on Testing Data:", lr_testing_data_accuracy)

Logistic Regression Accuracy on Testing Data: 0.8360655737704918


SVM Accuracy Score

In [40]:
#Accuracy on Training Data
svm_x_train_prediction = classifier.predict(x_train)
svm_training_data_accuracy = accuracy_score(svm_x_train_prediction,y_train)

In [78]:
print("SVM Accuracy on Training Data:", svm_training_data_accuracy)

SVM Accuracy on Training Data: 0.8553719008264463


In [59]:
#Accuraccy on test data
svm_x_test_prediction = classifier.predict(x_test)
svm_testing_data_accuracy = accuracy_score(svm_x_test_prediction,y_test)

In [79]:
print("SVM Accuracy on Testing Data:", svm_testing_data_accuracy)

SVM Accuracy on Testing Data: 0.819672131147541


DECISION TREE ACCURACY SCORE

In [62]:
#Accuracy on Training Data
dt_x_train_prediction = dt_model.predict(x_train)
dt_training_data_accuracy = accuracy_score(dt_x_train_prediction,y_train)

In [81]:
print("Decision Tree Accuracy on Training Data:", dt_training_data_accuracy)

Decision Tree Accuracy on Training Data: 1.0


In [64]:
#Accuraccy on test data
dt_x_test_prediction = dt_model.predict(x_test)
dt_testing_data_accuracy = accuracy_score(dt_x_test_prediction,y_test)

In [80]:
print("Decision Tree Accuracy on Testing Data:", dt_testing_data_accuracy)

Decision Tree Accuracy on Testing Data: 0.7868852459016393


RANDOM FOREST ACCURACY SCORE

In [46]:
#Accuracy on Training Data
rf_x_train_prediction = rf_model.predict(x_train)
rf_training_data_accuracy = accuracy_score(rf_x_train_prediction,y_train)

In [83]:
print(" Random Forest Accuracy on Training Data:", rf_training_data_accuracy)

 Random Forest Accuracy on Training Data: 1.0


In [48]:
#Accuraccy on test data
rf_x_test_prediction = rf_model.predict(x_test)
rf_testing_data_accuracy = accuracy_score(rf_x_test_prediction,y_test)

In [82]:
print("Random Forest Accuracy on Testing Data:", rf_testing_data_accuracy)

Random Forest Accuracy on Testing Data: 0.7704918032786885


Key Observations:<br>


*   Logistic Regression shows the best balance between training and testing accuracy, indicating good generalization.
*   SVM performs well but slightly less effectively on the test set.
*   Decision Tree and Random Forest models are overfitting the training data, which is causing them to underperform on the test data.

CROSS VALIDATION

In [75]:
# Cross-validation for Logistic Regression
cv_logistic = cross_val_score(logistic_model, x_train, y_train, cv=5)
print("Cross-validation scores for Logistic Regression:", cv_logistic)
print("Mean CV score for Logistic Regression:", cv_logistic.mean())

# Cross-validation for SVM
cv_svm = cross_val_score(classifier, x_train, y_train, cv=5)
print("Cross-validation scores for SVM:", cv_svm)
print("Mean CV score for SVM:", cv_svm.mean())

# Cross-validation for Random Forest
cv_rf = cross_val_score(rf_model, x_train, y_train, cv=5)
print("Cross-validation scores for Random Forest:", cv_rf)
print("Mean CV score for Random Forest:", cv_rf.mean())

# Cross-validation for Decision Tree
cv_dt = cross_val_score(dt_model, x_train, y_train, cv=5)
print("Cross-validation scores for Decision Tree:", cv_dt)
print("Mean CV score for Decision Tree:", cv_dt.mean())

/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:460: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:460: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _c

Cross-validation scores for Logistic Regression: [0.81632653 0.89795918 0.72916667 0.91666667 0.72916667]
Mean CV score for Logistic Regression: 0.8178571428571428
Cross-validation scores for SVM: [0.79591837 0.93877551 0.72916667 0.89583333 0.77083333]
Mean CV score for SVM: 0.8261054421768707
Cross-validation scores for Random Forest: [0.79591837 0.89795918 0.77083333 0.9375     0.75      ]
Mean CV score for Random Forest: 0.8304421768707483
Cross-validation scores for Decision Tree: [0.75510204 0.87755102 0.6875     0.8125     0.72916667]
Mean CV score for Decision Tree: 0.7723639455782313


Cross-Validation Results:<br>
Logistic Regression:

* Mean CV Score: 0.8179
* The model is fairly consistent but slightly underperforms compared to SVM and Random Forest.<br>

SVM (Support Vector Machine):

* Mean CV Score: 0.8261
* This model performs slightly better than Logistic Regression, showing strong generalization ability.
<br>

Random Forest:

* Mean CV Score: 0.8304
* The highest mean cross-validation score among the models, indicating it may be the best performer overall.
<br>

Decision Tree:

* Mean CV Score: 0.7724
* While still effective, the Decision Tree has a lower mean CV score compared to the other models, suggesting it may not generalize as well.

HYPERPARAMETER TUNING

In [93]:
# Hyperparameter tuning for Random Forest
param_grid_rf = {
    'n_estimators': [50, 100],
    'max_depth': [None, 10, 20],
    'min_samples_split': [2, 5],
    'min_samples_leaf': [1, 2]
}

grid_search_rf = GridSearchCV(estimator=rf_model, param_grid=param_grid_rf, cv=3, n_jobs=-1, verbose=2)
grid_search_rf.fit(x_train, y_train)
print("Best parameters for Random Forest:", grid_search_rf.best_params_)


Fitting 3 folds for each of 24 candidates, totalling 72 fits
Best parameters for Random Forest: {'max_depth': 10, 'min_samples_leaf': 2, 'min_samples_split': 2, 'n_estimators': 50}


In [89]:
# Hyperparameter tuning for Decision Tree
dt_params = {
    'max_depth': [None, 10, 20, 30],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4]
}

dt_grid_search = GridSearchCV(estimator=DecisionTreeClassifier(), param_grid=dt_params, cv=5, verbose=2, n_jobs=-1)
dt_grid_search.fit(x_train, y_train)
best_dt_model = dt_grid_search.best_estimator_

print("Best parameters for Decision Tree:", dt_grid_search.best_params_)
print("Best Decision Tree CV score:", dt_grid_search.best_score_)


Fitting 5 folds for each of 36 candidates, totalling 180 fits
Best parameters for Decision Tree: {'max_depth': 30, 'min_samples_leaf': 2, 'min_samples_split': 5}
Best Decision Tree CV score: 0.7886904761904763


In [92]:
# Hyperparameter tuning for SVM
param_grid_svm = {
    'C': [0.1, 1, 10],
    'kernel': ['linear', 'rbf']
}

grid_search_svm = GridSearchCV(estimator=classifier, param_grid=param_grid_svm, cv=3, n_jobs=-1, verbose=2)
grid_search_svm.fit(x_train, y_train)
print("Best parameters for SVM:", grid_search_svm.best_params_)

Fitting 3 folds for each of 6 candidates, totalling 18 fits
Best parameters for SVM: {'C': 0.1, 'kernel': 'linear'}


Best Parameters:<br>
Random Forest:<br>

* max_depth: 10
* min_samples_leaf: 2
* min_samples_split: 2
* n_estimators: 50


Decision Tree:

*  max_depth: 30
* min_samples_leaf: 2
* min_samples_split: 5
* Best CV Score: 0.7887


SVM:

* C: 0.1
* kernel: 'linear'

Re-train models with the best parameters

In [94]:
# Random Forest with best parameters
best_rf_model = RandomForestClassifier(
    max_depth=10,
    min_samples_leaf=2,
    min_samples_split=2,
    n_estimators=50
)
best_rf_model.fit(x_train, y_train)

RandomForestClassifier(max_depth=10, min_samples_leaf=2, n_estimators=50)

In [95]:
# Decision Tree with best parameters
best_dt_model = DecisionTreeClassifier(
    max_depth=30,
    min_samples_leaf=2,
    min_samples_split=5
)
best_dt_model.fit(x_train, y_train)

DecisionTreeClassifier(max_depth=30, min_samples_leaf=2, min_samples_split=5)

In [96]:
# SVM with best parameters
best_svm_model = svm.SVC(
    C=0.1,
    kernel='linear'
)
best_svm_model.fit(x_train, y_train)


SVC(C=0.1, kernel='linear')

EVALUATION

In [97]:
def evaluate_model(model, x_test, y_test):
    predictions = model.predict(x_test)
    accuracy = accuracy_score(predictions, y_test)
    return accuracy

# Evaluate Random Forest
rf_test_accuracy = evaluate_model(best_rf_model, x_test, y_test)
print("Random Forest Accuracy on Test Data:", rf_test_accuracy)

# Evaluate Decision Tree
dt_test_accuracy = evaluate_model(best_dt_model, x_test, y_test)
print("Decision Tree Accuracy on Test Data:", dt_test_accuracy)

# Evaluate SVM
svm_test_accuracy = evaluate_model(best_svm_model, x_test, y_test)
print("SVM Accuracy on Test Data:", svm_test_accuracy)

Random Forest Accuracy on Test Data: 0.819672131147541
Decision Tree Accuracy on Test Data: 0.7377049180327869
SVM Accuracy on Test Data: 0.8032786885245902


Analysis<br>
Performance Comparison:

* Random Forest achieved the highest accuracy among the models.
* SVM also performed well, closely following Random Forest.
* Decision Tree had the lowest accuracy, suggesting it might be overfitting or that its hyperparameters need further tuning.<br>

Model Selection:

* Based on accuracy, Random Forest is the best performing model for this dataset.


Build a Predictive System

In [98]:
input_data = (62,0,0,140,268,0,0,160,0,3.6,0,2,2)

# change the input data to a numpy array
input_data_as_numpy_array= np.asarray(input_data)

# reshape the numpy array as we are predicting for only on instance
input_data_reshaped = input_data_as_numpy_array.reshape(1,-1)

prediction = rf_model.predict(input_data_reshaped)
print(prediction)

if (prediction[0]== 0):
  print('The Person does not have a Heart Disease')
else:
  print('The Person has Heart Disease')

[0]
The Person does not have a Heart Disease


/usr/local/lib/python3.10/dist-packages/sklearn/base.py:465: UserWarning: X does not have valid feature names, but RandomForestClassifier was fitted with feature names
  warnings.warn(
